File: test/src/Snake.java
public class Snake{
    public Vec2 vec2;
    public SnakeBodySegement[] bodys;

    public Snake(Vec2 vec2,SnakeBodySegement[] bodys){
        this.vec2=vec2;
        this.bodys=bodys;
    }

    public static void main(String[] args) {
        //创建蛇头坐标
        Vec2 vec2 =new Vec2(20,20);
        //创建蛇的身体
        SnakeBodySegement bodys[] = {new SnakeBodySegement(Direction.down, 15),new SnakeBodySegement(Direction.right,20)};

        //创建蛇对象
        Snake snake = new Snake(vec2,bodys);
        System.out.println(snake.bodys);

        // for (int i = 0; i < 10; i++) {
        //     Vec2 vec =new Vec2(20+i,20+i);
        //     SnakeBodySegement bodys1[] = {new SnakeBodySegement(Direction.down, 15),new SnakeBodySegement(Direction.right,20)};
        //     bodys1[0]=new SnakeBodySegement(Direction.down, 15);
        //     bodys1[1]=new SnakeBodySegement(Direction.right,20);
        //     Snake snake1 = new Snake(vec,bodys1);
        // }
    }


    public static void drawSnake(Snake snake,int weight,int height){

    }
}

File: Test/src/Snake.ts
import { Vec2 } from "./Vec2"
export enum Direction {
    up,
    down,
    left,
    right,
}

export class SnakeBodySegment {
    /**
     * 身子节的方向。
     */
    public direction: Direction;
    public length: number;

    constructor(direction: Direction, length: number) {
        this.direction = direction;
        this.length = length;
    }
}

export class Snake {
    public head: Vec2;
    public body: SnakeBodySegment[];
    constructor(head: Vec2, body: SnakeBodySegment[]) {
        this.head = head;
        this.body = body;
    }
    /**
     * 行动一步
     */
    public step() {
        switch (this.body[0].direction) {
            case Direction.up:
                this.head.y++;
                break;
            case Direction.down:
                this.head.y--;
                break;
            case Direction.left:
                this.head.x++;
                break;
            case Direction.right:
                this.head.x--;
                break;
        }
        this.body[0].length ++;
        this.body[this.body.length-1].length --;
        if (this.body[this.body.length-1].length==0) {
            this.body.pop();
        }
    }
    /**
     * 转向  
     * 当它转向的时候，新增第一段，长度为1
     * 第二段长度-1
     * 调整蛇头位置
     */
    public turn(direction: Direction) {
        //当蛇第一段身体的方向为左或者右时候，左右移动跳出函数；
        if ((direction ==Direction.left || direction==Direction.right)
            && (this.body[0].direction == Direction.left || this.body[0].direction==Direction.right)) {
            return;
        }
        //当蛇第一段身体的方向为上或者下时候，上下移动跳出函数；
        if ((direction ==Direction.up || direction==Direction.down)
            && (this.body[0].direction == Direction.up || this.body[0].direction==Direction.down)) {
            return;
        }
        this.body.unshift(new SnakeBodySegment(direction,1));
        this.body[1].length--; 
        //改变蛇头位置
        switch (direction) {
            
            case Direction.up:
                this.head.y--;
                break;
            case Direction.down:
                this.head.y++;
                break;
            case Direction.left:
                this.head.x--;
                break;
            case Direction.right:
                this.head.x++;
                break;  
               
        }
    }
}